### DISCUSSION TITLE
Are there any dangers to using globally cached compiled statements ?

### DISCUSSION THREAD

**Discussion by thomas-mckay** [sqlalchemy/sqlalchemy#6966] [Usage Questions] [answered]
I've been migrating some of my projects from 1.3 to 1.4 and got quite interested in the new statement caching feature of 1.4. I ran a few benchmarks to gauge the speed advantages of implementing lambdas in my code:

```python
from timeit import timeit

import sqlalchemy as sa
from sqlalchemy import create_engine
from sqlalchemy import select


metadata = sa.MetaData()

kv_pair_tbl = sa.Table(
    'kv_store',
    metadata,
    sa.Column('key', sa.String(64), primary_key=True),
    sa.Column('value', sa.String(64), nullable=False),
)

engine = create_engine('sqlite://')
metadata.create_all(engine)

with engine.begin() as conn:
    conn.execute(sa.insert(kv_pair_tbl).values(dict(key='foo', value='bar')))


def get_regular_caching():
    stmt = (
        select(kv_pair_tbl.c.value)
        .where(kv_pair_tbl.c.key == sa.bindparam('key'))
    )
    return engine.execute(stmt, key='foo').one()


def get_lambda_caching():
    stmt = (
        sa.lambda_stmt(lambda: (
            select(kv_pair_tbl.c.value)
            .where(kv_pair_tbl.c.key == sa.bindparam('key'))
        ))
    )
    return engine.execute(stmt, key='foo').one()


compiled_stmt = None


def get_custom_caching():
    global compiled_stmt

    if compiled_stmt is None:
        compiled_stmt = (
            select(kv_pair_tbl.c.value)
            .where(kv_pair_tbl.c.key == sa.bindparam('key'))
        ).compile(dialect=engine.dialect)

    return engine.execute(compiled_stmt, key='foo').one()

def get_raw():
    return engine.execute("SELECT key, value FROM kv_store WHERE key=:key", key='foo').one()


print(timeit("get_regular_caching()", number=10_000, globals=globals()))  # 1.71  (1.97 on v1.3.10)
print(timeit("get_lambda_caching()", number=10_000, globals=globals()))  # 1.00
print(timeit("get_custom_caching()", number=10_000, globals=globals()))  # 0.60
print(timeit("get_raw()", number=10_000, globals=globals()))  # 0.65
```

I ran this multiple times and kept the best figures of each run.

With a simple select, 1.4.22 (with de-facto query caching on the engine, if I understood the docs correctly) performed 10% better than 1.3.10. Using the lambda form gave 40% better performance on that figure. Finally, out of curiosity, and because there was still a large gap between lambdas and a raw call, I started to suspect that maybe generating the cache keys from the lambda also had a non-negligible overhead. So I did a test with a pre-compiled, globally cached statement and that gave me 40% better perfs than the lambdas (3.3x faster than a regular select on 1.3.10).

So my question is: **is it ill-advised/unsafe to keep some queries declared pre-compiled in a module (or any global object, really) and to import/get them as needed?** Please, also consider threaded/async execution environment in your answer. I've looked at the SQLAlchemy sources (specifically `sqlalchemy.sql.compiler.py` and the connection/engine code that interacts with these objects and I haven't seen anything like mutating states that would make them dangerous to pass around. Am I missing something ?

Either way, kudos to the dev team for the feature, as it seems to improve performance greatly for small resultset. I wasn't able to reproduce the percentages shown in the documentation, but still 50% better perfs on lambdas vs 1.3.10 is nothing to sneeze at.

**Comment by CaselIT** [comment]
I'll leave the prover answer to mike.

Note that if the queries are static and use bindparams, just can just move the select at module level. Doing
```py
g_stmt = select(kv_pair_tbl.c.value).where(
    kv_pair_tbl.c.key == sa.bindparam("key")
)

def get_global_select():
    return engine.execute(g_stmt, key="foo").one()
```
results in
```
get_regular_caching 2.1977678
get_lambda_caching 1.1311661000000002
get_custom_caching 0.6412933000000001
get_raw 0.7121718000000001
get_global_select 0.7973981000000006
```

So almost the same as the raw query.

Second note: engine.execute is deprecated, you should use a connection

**Reply by thomas-mckay** [reply]
@CaselIT I went with `bindparam` for the example by accident but if I think about it, most of the times it would fit my use case.

Considering your results, would it be fair to assume that the majority of the overhead in this example comes from building the query (as opposed to compiling it)? This might explain why I don't quite reach the performances stated in the docs in my tests.

> Second note: engine.execute is deprecated, you should use a connection

Thanks, I completely missed that deprecation.

**Reply by CaselIT** [reply]
> Considering your results, would it be fair to assume that the majority of the overhead in this example comes from building the query (as opposed to compiling it)?

for a simple query what would be the case, since when creating the query the select needs to coerce all the attributes to verity that they make sense, and that's a somewhat expensive thing to do. The compilation in 1.4 once done is cached, so it's overhead is the generation of the cache key, that's not free, but it should still be better than not having it :)

> Thanks, I completely missed that deprecation.

you can use `python -Werror file.py` to surface all warnings as errors

**Reply by thomas-mckay** [reply]
> you can use python -Werror file.py to surface all warnings as errors

I did but it didn't error (no deprecation warning is shown either with a regular `python myfile.py`) (on 1.4.23, python is 3.9.6, btw)

**Reply by CaselIT** [reply]
you are right, this is behind the `SQLALCHEMY_WARN_20` flag as explained here https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-to-2-0-step-two-turn-on-removedin20warnings

**Comment by zzzeek** [comment, marked as answer]
first lets make this test a better test.   don't use engine.execute().  there is a huge amount of overhead associated with that, checking out from the pool, starting a transaction, lets do all of that outside of the tests.  lets also show a real baseline of 1.4 *without* caching to get a clear picture of what's happening:

```py
from timeit import timeit

import sqlalchemy as sa
from sqlalchemy import create_engine
from sqlalchemy import select

sqla_13 = sa.__version__.startswith("1.3")

metadata = sa.MetaData()

kv_pair_tbl = sa.Table(
    "kv_store",
    metadata,
    sa.Column("key", sa.String(64), primary_key=True),
    sa.Column("value", sa.String(64), nullable=False),
)

engine = create_engine("sqlite://")
metadata.create_all(engine)

with engine.begin() as conn:
    conn.execute(sa.insert(kv_pair_tbl).values(dict(key="foo", value="bar")))

connection = engine.connect()
connection.begin()
connection_no_cache = engine.connect().execution_options(compiled_cache=None)
connection_no_cache.begin()


def get_no_caching():
    stmt = select(kv_pair_tbl.c.value).where(
        kv_pair_tbl.c.key == sa.bindparam("key")
    )
    return connection_no_cache.execute(stmt, key="foo").fetchone()


def get_regular_call():
    stmt = select([kv_pair_tbl.c.value]).where(
        kv_pair_tbl.c.key == sa.bindparam("key")
    )
    return connection.execute(stmt, key="foo").fetchone()


def get_lambda_caching():
    stmt = sa.lambda_stmt(
        lambda: (
            select(kv_pair_tbl.c.value).where(
                kv_pair_tbl.c.key == sa.bindparam("key")
            )
        )
    )
    return connection.execute(stmt, key="foo").fetchone()


the_stmt = None


def get_just_reuse_the_stmt():
    global the_stmt
    if the_stmt is None:
        the_stmt = select([kv_pair_tbl.c.value]).where(
            kv_pair_tbl.c.key == sa.bindparam("key")
        )
    return connection.execute(the_stmt, key="foo").fetchone()


compiled_stmt = None


def get_custom_caching():
    global compiled_stmt

    if compiled_stmt is None:
        compiled_stmt = (
            select([kv_pair_tbl.c.value]).where(
                kv_pair_tbl.c.key == sa.bindparam("key")
            )
        ).compile(dialect=engine.dialect)

    return connection.execute(compiled_stmt, key="foo").fetchone()


def get_raw():
    return connection.execute(
        "SELECT key, value FROM kv_store WHERE key=:key", key="foo"
    ).fetchone()


def _timeit(fn_name, **kw):
    return f"{fn_name} : {timeit(fn_name, **kw)}"


if sqla_13:
    print(_timeit("get_regular_call()", number=10_000, globals=globals()))
    print(
        _timeit("get_just_reuse_the_stmt()", number=10_000, globals=globals())
    )
    print(_timeit("get_custom_caching()", number=10_000, globals=globals()))
    print(_timeit("get_raw()", number=10_000, globals=globals()))
else:
    print(_timeit("get_no_caching()", number=10_000, globals=globals()))
    print(_timeit("get_regular_call()", number=10_000, globals=globals()))
    print(_timeit("get_lambda_caching()", number=10_000, globals=globals()))
    print(
        _timeit("get_just_reuse_the_stmt()", number=10_000, globals=globals())
    )
    print(_timeit("get_custom_caching()", number=10_000, globals=globals()))
    print(_timeit("get_raw()", number=10_000, globals=globals()))
```

on my machine here is 1.4:

```
get_no_caching() : 1.7397529390000273
get_regular_call() : 1.0548595329892123
get_lambda_caching() : 0.48266025198972784
get_just_reuse_the_stmt() : 0.3074115880008321
get_custom_caching() : 0.20178842499444727
get_raw() : 0.2669021990004694
```

here is 1.3:

```
get_regular_call() : 1.279716165998252
get_just_reuse_the_stmt() : 0.7238781789928908
get_custom_caching() : 0.22416703299677465
get_raw() : 0.23468372500792611

```

so here's what we can observe:

1. 1.4's caching grants a 38% speedup vs . non cached
2. 1.4's compilation is slower than that of 1.3.  this is because it does more things such as [from linting](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#built-in-from-linting-will-warn-for-any-potential-cartesian-products-in-a-select-statement)
3. there's not any need to call .compile() if holding onto a global statement.  just use the same statement object, the building of it (expensive) and construction of the cache key (expensive) is only done once.   1.3 actually has a compiled cache that can leverage this idea also if you use the same statement object each time.
4. the real place that 1.4's caching shines is that **the ORM compilation is also cached**, which is a dramatically more expensive task than core compilation.   The performance suite at https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.performance should be considered the canonical source of comparison of this kind of thing.
5. the additional expense that lambda caching fixes is to some extent the cache key generation but the most expensive thing right now is the [coercions](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/sql/coercions.py) of all the arguments.  the coercions are how we have nice error messages and processing of incoming expressions, but there's expense with that, which might be a good candidate for cython later on.


all of that said, if global statements are an option then you can of course use the statement globally.   that or holding onto the compiled object globally is perfectly fine there are no mutations to these things once they are generated.

**Reply by thomas-mckay** [reply]
Thank you for the confirmation and for the detailed answer, it's very informative, as always. Lambda caching seems like a good place for me to start getting better perfs out of some simple but recurrent queries, indeed. It's also good to know that if the need arises, I can use global statements. Thanks :)
